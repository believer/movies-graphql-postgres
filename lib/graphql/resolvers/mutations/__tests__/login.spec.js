const login = require('../login')
const { invalidPassword } = require('../../../../services/login')
const { issueToken } = require('../../../../services/token')

jest.mock('../../../../services/login', () => ({
  invalidPassword: jest.fn(),
}))

jest.mock('../../../../services/token', () => ({
  issueToken: jest.fn().mockReturnValue('token'),
}))

describe('mutations/login', () => {
  let db

  beforeEach(() => {
    db = {
      one: jest.fn(),
    }

    issueToken.mockClear()
  })

  it('should get a specific user', async () => {
    try {
      db.one.mockReturnValue({
        name: 'Bruce Wayne',
        id: 3,
        password: 'p4zzw0rd',
      })

      await login(
        null,
        { input: { username: 'batman@waynecorp.org', password: '' } },
        { db }
      )

      expect(db.one.mock.calls[0][0]).toMatchSnapshot()
    } catch (e) {
      throw new Error(e)
    }
  })

  it('should return error if wrong password', async () => {
    db.one.mockImplementation(() =>
      Promise.resolve({ password: 'hej', username: 'batman' })
    )
    invalidPassword.mockReturnValue(true)

    try {
      await login(
        null,
        {
          input: { username: 'batman@waynecorp.org', password: 'robin4ever' },
        },
        { db }
      )
    } catch (e) {
      expect(e.message).toMatch(/Wrong username or password/)
    }
  })

  it('should issue a token', async () => {
    try {
      const user = {
        name: 'Bruce Wayne',
        email: 'batman@waynecorp.org',
        id: 3,
        password: 'test',
        salt: 'salt',
      }
      const params = {
        input: { username: 'batman@waynecorp.org', password: '' },
      }

      invalidPassword.mockReturnValue(false)
      db.one.mockReturnValue(user)

      await login(null, params, { db })

      expect(issueToken.mock.calls[0][0]).toMatchSnapshot()
    } catch (e) {
      throw new Error(e)
    }
  })

  it('should return database data', async () => {
    try {
      invalidPassword.mockReturnValue(false)
      db.one.mockReturnValue({
        name: 'Bruce Wayne',
        id: 3,
        password: 'p4zzw0rd',
      })

      const response = await login(
        null,
        {
          input: { username: 'batman@waynecorp.org', password: '' },
        },
        { db }
      )

      expect(response).toEqual({
        token: 'token',
      })
    } catch (e) {
      throw new Error(e)
    }
  })

  it('should handle errors', async () => {
    expect.hasAssertions()
    db.one.mockImplementation(() => Promise.reject('nope'))

    try {
      await login(
        null,
        {
          input: { username: 'batman@waynecorp.org', password: '' },
        },
        { db }
      )
    } catch (e) {
      expect(e.message).toMatch(/nope/)
    }
  })
})
