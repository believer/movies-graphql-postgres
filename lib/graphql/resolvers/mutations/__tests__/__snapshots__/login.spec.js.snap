// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutations/login should get a specific user 1`] = `"SELECT * FROM users WHERE email = 'batman@waynecorp.org'"`;

exports[`mutations/login should issue a token 1`] = `
Object {
  "email": "batman@waynecorp.org",
  "id": 3,
  "name": "Bruce Wayne",
}
`;
